src/basic/virt.c

```diff
@@ -460,11 +460,30 @@ int detect_container(void) {
         _cleanup_free_ char *o = NULL;
         _cleanup_free_ char *p = NULL;
         const char *e = NULL;
+        unsigned i;
         int r;
 
+        static const struct {
+                const char *file_path;
+                int id;
+        } container_file_table[] = {
+                /* https://github.com/moby/moby/issues/18355 */
+                { "/.dockerenv",        VIRTUALIZATION_DOCKER },
+                /* https://github.com/containers/podman/issues/6192 */
+                /* https://github.com/containers/podman/issues/3586#issuecomment-661918679 */
+                { "/run/.containerenv", VIRTUALIZATION_PODMAN },
+        };
+
         if (cached_found >= 0)
                 return cached_found;
 
+        for (i = 0; i < ELEMENTSOF(container_file_table); i++) {
+                if (access(container_file_table[i].file_path, F_OK) >= 0) {
```

Comment:
i'd prefer if we'd generate a debug message if things fail because of any other error than ENOENT here. i.e. do: 

```c
if (errno != ENOENT)
        log_debug_errno(errno, "Checking if %s exists failed, ignoring: %m");
```

that makes things more debuggable, i.e. track down perms issues and stuff.

(yes, the old code doesn't do this right either, but for new code we strive to be better and catch errors like this at least with a debug msg)

Reply:
(I fixed up some of the old code now here: #18209 — would love a review...)